query_database.py

```diff
@@ -0,0 +1,405 @@
+    'Finnish Museum of Natural History': 'FMNH',
+    'Naturalis Biodiversity Center': 'Naturalis'
+}
+
+
+def config(filename='database.ini', section='postgresql'):
+    """ Sets up the basic database connection rules fur further usage
+        :return: db: instance of the database's properties
+    """
+
+    # create a parser
+    parser = ConfigParser()
+    # read config file
+    parser.read(filename)
+
+    # get section, default to postgresql
+    db = {}
+    if parser.has_section(section):
+        params = parser.items(section)
+        for param in params:
+            db[param[0]] = param[1]
+
+    return db
+
+
+def insert_countries(country_entity) -> str:
+    columns: str = ''
+    values: str = ''
+
+    for key in country_entity.keys():
+        if len(columns) == 0:
+            columns += key
+        else:
+            columns += ', ' + key
+
+    for value in country_entity.values():
+        if len(values) == 0:
+            values += "'" + str(value) + "'"
+        else:
+            values += ", '" + str(value) + "'"
+
+    query = f"INSERT INTO countries ({columns}) VALUES ({values})"
+
+    return query
+
+
+def update_countries(country_entity, row) -> str:
+    column_values: str = ''
+
+    for key in country_entity:
+        column_values += f"{key} = '{country_entity[key]}', "
+
+    query = f"UPDATE countries SET {column_values[:-2]} WHERE id = {row[0]}"
+
+    return query
+
+
+def insert_countries_data(gbif_datasets, gbif_specimens, gbif_issues_flags, geocase_data):
+    """ Transforms the received countries' data to a standardised format
+        Saves the formatted data in the database by insert or update
+    """
+
+    # Temporary mapping between GBIF and GeoCASe
+    country_mapping = {
+        'DE': 'Germany',
+        'GB': 'UK',
+        'EE': 'Estonia',
+        'AT': 'Austria',
+        'FI': 'Finland',
+        'NL': 'The Netherlands'
+    }
+
+    # For each organisation, prepare and save data to database
+    for country in gbif_datasets['countries']:
+        # Check if GeoCASe data from country is present
+        geocase_dummy: dict = {}
+
+        if country in country_mapping:
+            geocase_dummy = geocase_data['countries'][country_mapping[country]]
+
+        # Initiate basic entity for database
+        country_entity = {
+            'country_code': country,
+            'country_name': 'test',
+            'datasets_count': json.dumps({
+                'gbif': gbif_datasets['countries'][country]
+            }),
+            'specimens_count': json.dumps({
+                'gbif': gbif_specimens['countries'][country],
+                'geocase': geocase_dummy
+            }),
+            'issues_flags': json.dumps(gbif_issues_flags['countries'][country])
+        }
+
+        # Create database connection
+        params = config()
+        conn = psycopg2.connect(**params)
+        cur = conn.cursor()
+
+        # Check if entity already exists (for same month)
```

Comment:
This is also possible with an upsert query:
https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/

Reply:
Combined separate insert and update functions (queries) to one upset query for both ways